File: src/main/java/org/lanternpowered/server/effect/entity/particle/item/ItemDeathParticleEffect.java
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.effect.entity.particle.item;

import com.flowpowered.math.vector.Vector3d;
import org.lanternpowered.server.effect.entity.EntityEffect;
import org.lanternpowered.server.entity.LanternEntity;
import org.spongepowered.api.effect.particle.ParticleEffect;
import org.spongepowered.api.effect.particle.ParticleTypes;
import org.spongepowered.api.util.AABB;

public class ItemDeathParticleEffect implements EntityEffect {

    static final class EffectHolder {

        static final ParticleEffect DEATH_EFFECT =
                ParticleEffect.builder().type(ParticleTypes.CLOUD).quantity(3).offset(Vector3d.ONE.mul(0.1)).build();
    }

    @Override
    public void play(LanternEntity entity) {
        entity.getWorld().spawnParticles(EffectHolder.DEATH_EFFECT,
                entity.getBoundingBox().map(AABB::getCenter).orElseGet(entity::getPosition));
    }
}

File: src/main/kotlin/org/lanternpowered/server/effect/entity/particle/item/ItemDeathParticleEffect.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.effect.entity.particle.item

import org.lanternpowered.server.effect.entity.EntityEffect
import org.lanternpowered.server.entity.LanternEntity
import org.spongepowered.api.effect.particle.ParticleEffect
import org.spongepowered.api.effect.particle.ParticleTypes
import org.spongepowered.math.vector.Vector3d

object ItemDeathParticleEffect : EntityEffect {

    internal object EffectHolder {
        
        val DEATH_EFFECT: ParticleEffect = ParticleEffect.builder()
                .type(ParticleTypes.CLOUD).quantity(3).offset(Vector3d.ONE.mul(0.1)).build()
    }

    override fun play(entity: LanternEntity) {
        /*
        entity.world.spawnParticles(EffectHolder.DEATH_EFFECT,
                entity.boundingBox.map { obj: AABB -> obj.center }.orElseGet { entity.position })*/
        // TODO
    }
}
